adds possibility to include user defined dict in serialized nodes

abak opened the pull request:
In my use case, I need additional data to be serialized with the workflow dag's nodes. 
This PR introduces a dict member to `WorkflowNode`, `user_params`, that can be used to convey such data.

I added tests to make sure that the new member is correctly serialized, and took care to not break retrocompatibility with existing serialized data.

This MR addresses issue #3

Commit:
adds possibility to include user defined dict in serialized nodes

brabiega reviewed (commented):
Could you also update Readme.md file describing how to use this mechanism?

Commit:
adds same thing for Workflow

Commit:
update README and add example

abak commented:
@brabiega even though i do not use it, i added the same thing for the Workflow instance, and updated the README as requested.

abak commented:
I force pushed following a rebase on master, to take the 0.7.0 release into account.

brabiega commented on README.md:
```diff
@@ -123,6 +123,24 @@ Handler params are:
 
+
+### Support for custom data
+Both `Workflow`and `WorkflowNode` have a `user_params` dictionary member that can be used to store 
```
Would you mind if I call it ```custom_payload```?
```user_params``` indicates that it should store some user parameters, which brings more questions - e.g. what kind of params? At least that's my reasoning.

abak replied:
oh yeah that is definetely a much better name !

brabiega commented on celery_dyrygent/workflows/node.py:
```diff
@@ -46,12 +47,17 @@ def to_dict(self):
             signature=self.signature,
             dependencies=self.dependencies,
         )
+        if not self.user_params :
```
I don't think there is a need for this ```if```.
Let's always return dict with ```user_params``` or ```custom_payload``` however we decide to call it.

abak replied:
Fine. I only wanted to make sure that object that were serialized with versions up to 0.7.0 could still be deserialized, but i'll modify those checks as you suggested.

brabiega replied:
No worries. It works. I just tested with wf published using 0.6.1 and processed with 0.7.0 + your changes. Works flawlessly

brabiega commented on celery_dyrygent/workflows/node.py:
```diff
@@ -62,4 +68,8 @@ def from_dict(cls, data_dict):
         sig = entities.Signature(data_dict['signature'])
         obj = cls(sig)
         obj.dependencies = data_dict['dependencies']
+        if 'user_params' in data_dict:
```
Let's simplify
```
obj.user_params = data_dict.get('user_params', {})
```

abak replied:
:+1:

brabiega commented on celery_dyrygent/workflows/workflow.py:
```diff
@@ -519,6 +521,10 @@ def to_dict(self):
             res[attr] = getattr(self, attr)
         res['stats'] = self.stats
+
+        if self.user_params:
```
Again I don't think this ```if``` is needed.
Simpler code = better code

brabiega commented on celery_dyrygent/workflows/workflow.py:
```diff
@@ -529,6 +535,8 @@ def from_dict(cls, workflow_dict):
             for node_id, node_dict in workflow_dict['nodes'].items()
         }
 
+        obj.user_params = workflow_dict['user_params'] if 'user_params' in workflow_dict else {}
```
```
obj.user_params = workflow_dict.get('user_params', {})
```

brabiega commented:
Sorry it took me so long to take a look :)

Commit:
Code Review :

- Rename added dict from user_params to custom_payload
- simplify ser/deser by removing always including custom_payload
- replace ifs by getter with default

brabiega commented on tests/unit/test_workflows.py:
```diff
@@ -101,12 +101,15 @@ def test_to_dict(self, sig_factory):
         assert d['dependencies']
         assert d['dependencies']['task-2'] is None
         assert d['dependencies']['task-3'] is None
+        assert d['custom_payload']['foo'] is "bar"
```
I know it works, but it looks awkward :)
```
assert d['custom_payload']['foo'] == "bar"
```

abak replied:
No problem to fix it, but just fyi, it's used several times throughout that file (e.g. lines 113, 166, 167)

brabiega commented on tests/unit/test_workflows.py:
```diff
@@ -548,10 +571,41 @@ def test_to_dict(self, wf):
+
+        wf.custom_payload["foo"] = "bar"
+        res = wf.to_dict()
+        assert res == {
```
This test is a kind of redundant. 
I'd like to keep the tests clean and practical.

```
    def test_to_dict_with_custom_payload(self, wf):
        wf.custom_payload["foo"] = "bar"
        res = wf.to_dict()
        assert res['custom_payload']['foo'] == 'bar'
```
This should be good enough.

abak replied:
sure thing

brabiega reviewed (changes requested):
Just a few minor changes in tests please. Once that's done I'll merge it :+1:

Commit:
Tests adjustments

brabiega reviewed (approved):
Alright. Looking good
